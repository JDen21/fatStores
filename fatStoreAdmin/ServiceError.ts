import type { possibleErrorTypes } from "./types.d.ts";

export default class ServiceError extends Error {
  devMessage: string = "";
  code: number = 0;
  rootError?: Error;
  private static defaultError = "Unknown error.";

  constructor(
    code: number = 0,
    externalServiceError: possibleErrorTypes,
    message: string,
  ) {
    super();
    this.message = message;
    this.code = code;

    // * assume external service error is not known
    if (externalServiceError === null) {
      this.devMessage = ServiceError.defaultError;
    }

    // * save error obj
    if (externalServiceError instanceof Error) {
      this.rootError = externalServiceError;
      this.devMessage = externalServiceError.message;
    }

    // * assume external service hints error code
    // * this code is sent as response status so
    // * only use if uncertain of proper error code.
    if (typeof externalServiceError === "number" && this.code === 0) {
      this.code = externalServiceError;
      this.devMessage = ServiceError.defaultError;
    }

    // * assumes external service error is a simple message
    if (typeof externalServiceError === "string") {
      this.devMessage = externalServiceError;
    }

    // * default to Bad Gateway if nothing is certain
    if (this.code === 0) {
      this.code = 502;
    }
  }
}
