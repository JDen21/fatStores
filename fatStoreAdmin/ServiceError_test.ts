import { assertEquals, assertStrictEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import ServiceError from "./ServiceError.ts";

const runSnapTest = Deno.args.includes("--snap") === false;

Deno.test({
  name: "ServiceError/UnknownNullError",
  fn: function () {
    const responseMessage = "Error response to request";
    const unknownErrorMes = "Unknown error.";

    const unknownNullError = new ServiceError(500, null, responseMessage);
    assertEquals(unknownNullError.code, 500, "should save error code");
    assertEquals(
      unknownNullError.devMessage,
      unknownErrorMes,
      "should save message to developer",
    );
    assertEquals(
      unknownNullError.message,
      responseMessage,
      "should save error response to request",
    );
  },
});

Deno.test({
  name: "ServiceError/InheritedError",
  fn: function () {
    const responseMessage = "Error response to request";
    const externalErrorDevMes = "Error from external";

    const externError = new Error(externalErrorDevMes);
    const inheritedError = new ServiceError(500, externError, responseMessage);
    assertEquals(inheritedError.code, 500, "should save error code");
    assertEquals(
      inheritedError.devMessage,
      externalErrorDevMes,
      "should save message to developer",
    );
    assertEquals(
      inheritedError.message,
      responseMessage,
      "should save error response to request",
    );
    assertStrictEquals(inheritedError.rootError, externError);
  },
});

Deno.test({
  name: "ServiceError/MessageError",
  fn: function () {
    const responseMessage = "Error response to request";
    const unknownErrorMes = "Unknown error.";

    // * keep message as log to developer but dont forward to response.
    const messageError = new ServiceError(
      500,
      unknownErrorMes,
      responseMessage,
    );
    assertEquals(messageError.code, 500, "should save error code");
    assertEquals(
      messageError.devMessage,
      unknownErrorMes,
      "should save message to developer",
    );
    assertEquals(
      messageError.message,
      responseMessage,
      "should save error response to request",
    );
  },
});

Deno.test({
  name: "ServiceError/CodeError",
  fn: function () {
    const responseMessage = "Error response to request";
    const unknownErrorMes = "Unknown error.";

    // * let the external service decide which error code to respond.
    const codeError = new ServiceError(0, 512, responseMessage);
    assertEquals(codeError.code, 512, "should save error code");
    assertEquals(
      codeError.devMessage,
      unknownErrorMes,
      "should save message to developer",
    );
    assertEquals(
      codeError.message,
      responseMessage,
      "should save error response to request",
    );
  },
});

Deno.test({
  name: "ServiceError/DefaultGatewayError",
  fn: function () {
    const responseMessage = "Error response to request";
    const unknownErrorMes = "Unknown error.";

    const defaultGatewayError = new ServiceError(
      0,
      unknownErrorMes,
      responseMessage,
    );
    assertEquals(defaultGatewayError.code, 502, "should save error code");
    assertEquals(
      defaultGatewayError.devMessage,
      unknownErrorMes,
      "should save message to developer",
    );
    assertEquals(
      defaultGatewayError.message,
      responseMessage,
      "should save error response to request",
    );
  },
});

Deno.test({
  name: "ServiceError snaps/UnknownNullError",
  ignore: runSnapTest,
  fn: async (t) => {
    const responseMessage = "Error response to request";
    const unknownNullError = new ServiceError(500, null, responseMessage);
    await assertSnapshot(t, JSON.stringify(unknownNullError));
  },
});

Deno.test({
  name: "ServiceError snaps/InheritedError",
  ignore: runSnapTest,
  fn: async (t) => {
    const externError = new Error("Error from external");
    const inheritedError = new ServiceError(
      500,
      externError,
      "Error response to request",
    );
    await assertSnapshot(t, JSON.stringify(inheritedError));
  },
});

Deno.test({
  name: "ServiceError snaps/MessageError",
  ignore: runSnapTest,
  fn: async (t) => {
    const messageError = new ServiceError(
      500,
      "Unknown error.",
      "Error response to request",
    );
    await assertSnapshot(t, JSON.stringify(messageError));
  },
});

Deno.test({
  name: "ServiceError snaps/CodeError",
  ignore: runSnapTest,
  fn: async (t) => {
    const responseMessage = "Error response to request";

    // * let the external service decide which error code to respond.
    const codeError = new ServiceError(0, 512, responseMessage);
    await assertSnapshot(t, JSON.stringify(codeError));
  },
});

Deno.test({
  name: "ServiceError snaps/DefaultGatewayError",
  ignore: runSnapTest,
  fn: async (t) => {
    const responseMessage = "Error response to request";
    const unknownErrorMes = "Unknown error.";

    const defaultGatewayError = new ServiceError(
      0,
      unknownErrorMes,
      responseMessage,
    );
    await assertSnapshot(t, JSON.stringify(defaultGatewayError));
  },
});
