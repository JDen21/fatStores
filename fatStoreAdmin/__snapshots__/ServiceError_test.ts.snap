export const snapshot = {};

snapshot[`ServiceError snaps/UnknownNullError 1`] = `'{"devMessage":"Unknown error.","code":500,"message":"Error response to request"}'`;

snapshot[`ServiceError snaps/InheritedError 1`] = `'{"devMessage":"Error from external","code":500,"rootError":{},"message":"Error response to request"}'`;

snapshot[`ServiceError snaps/MessageError 1`] = `'{"devMessage":"Unknown error.","code":500,"message":"Error response to request"}'`;

snapshot[`ServiceError snaps/CodeError 1`] = `'{"devMessage":"Unknown error.","code":512,"message":"Error response to request"}'`;

snapshot[`ServiceError snaps/DefaultGatewayError 1`] = `'{"devMessage":"Unknown error.","code":502,"message":"Error response to request"}'`;
